src/appleseed/renderer/modeling/phasefunction/henyeyphasefunction.cpp

```diff
@@ -0,0 +1,359 @@
+    }
+
+    virtual void prepare_inputs(
+        Arena&                arena,
+        const ShadingRay&     volume_ray,
+        void*                 data
+        ) const APPLESEED_OVERRIDE
+    {
+        InputValues* values = static_cast<InputValues*>(data);
+
+        // Precompute extinction.
+        values->m_precomputed.m_normalized_extinction =
+            values->m_absorption_multiplier * values->m_absorption +
+            values->m_scattering_multiplier * values->m_scattering;
+
+        // Ensure that extinction spectrum has unit norm, which is neccessary for distance sampling.
+        const float extinction_norm = max_value(values->m_precomputed.m_normalized_extinction);
+        values->m_precomputed.m_normalized_extinction /= extinction_norm;
+        values->m_precomputed.m_extinction_multiplier *= extinction_norm;
+    }
+
+    virtual float sample_distance(
+        SamplingContext&       sampling_context,
+        const ShadingRay&      volume_ray,
+        const void*            data,
+        float&                 distance
+        ) const APPLESEED_OVERRIDE
+    {
+        const InputValues* values = static_cast<const InputValues*>(data);
+
+        const float ray_length = static_cast<float>(
+            norm(volume_ray.m_dir) *
+            (volume_ray.m_tmax - volume_ray.m_tmin));
+
+        // Sample distance.
+        sampling_context.split_in_place(1, 1);
+        const float s = sampling_context.next2<float>();
+        distance = sample_exponential_distribution_on_segment(
+            s, values->m_precomputed.m_extinction_multiplier, 0.0f, ray_length);
+
+        // Return corresponding PDF value.
+        return exponential_distribution_on_segment_pdf(
+            distance, values->m_precomputed.m_extinction_multiplier, 0.0f, ray_length);
+    }
+
+    virtual float sample(
+        SamplingContext&      sampling_context,
+        const ShadingRay&     volume_ray,
+        const void*           data,
+        float                 distance,
+        Vector3f&             incoming) const APPLESEED_OVERRIDE
+    {
+        const InputValues* values = static_cast<const InputValues*>(data);
+        const Vector3f outgoing = Vector3f(normalize(volume_ray.m_dir));
+        const Basis3f basis(outgoing);
+
+        //
+        // x = 1/(2g) * (1 + g^2 - [(1 - g^2) / (1 + g*s)]^2),
+        // where x is cos(phi) and s is uniform random sample from [-1; 1].
+        //
+
+        const float g = values->m_average_cosine;
+        const float sqr_g = g * g;
+
+        sampling_context.split_in_place(1, 2);
+        const float s = 2.0f * sampling_context.next2<float>() - 1.0f;
+
+        float cosine;
+        if (std::abs(g) < +1e-5f)
+        {
+            cosine = s; // isotropic
+        }
+        else
+        {
+            const float t = (1.0f - sqr_g) / (1.0f + g*s);
+            cosine = 0.5f / g * (1.0f + sqr_g - t * t);
+        }
+        const float sine = std::sqrt(1.0f - cosine * cosine);
```

Comment:
You might want to clamp cosine^2 to 1 since any numerical inaccuracy will lead to NaN otherwise.